pkg/webhook/resources/setting/validator.go

```diff
@@ -709,6 +723,133 @@ func (v *settingValidator) validateDeleteStorageNetwork(setting *v1beta1.Setting
 	return werror.NewMethodNotAllowed(fmt.Sprintf("Disallow delete setting name %s", settings.StorageNetworkName))
 }
 
+func (v *settingValidator) findVolumesByLabelAppName(namespace string, labelAppNameValue string) ([]string, error) {
+	volumes := make([]string, 0)
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValue,
+	}
+
+	pvcs, err := v.pvcCache.List(namespace, sets.AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pvc := range pvcs {
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+		logrus.Debugf("Get system volume %v", pvc.Spec.VolumeName)
+		volumes = append(volumes, pvc.Spec.VolumeName)
+	}
+
+	return volumes, nil
+}
+
+func (v *settingValidator) getPrometheusVolumes() ([]string, error) {
+	return v.findVolumesByLabelAppName(util.CattleMonitoringSystemNamespace, labelAppNameValuePrometheus)
+}
+
+func (v *settingValidator) getAlertManagerVolumes() ([]string, error) {
+	return v.findVolumesByLabelAppName(util.CattleMonitoringSystemNamespace, labelAppNameValueAlertManager)
+}
+
+func (v *settingValidator) getGrafanaVolume() (string, error) {
+	volumes, err := v.findVolumesByLabelAppName(util.CattleMonitoringSystemNamespace, labelAppNameValueGrafana)
+	if err != nil {
+		return "", err
+	}
+
+	if len(volumes) == 0 {
+		return "", nil
+	}
+
+	//grafana deployment only has one pvc
```

Comment:
I'd suggest not assuming this. Considering the whole returned slice makes the code more robust.

Reply:
Refactor per suggestion, thanks.